clippy_lints/src/methods/join_absolute_paths.rs

```diff
@@ -0,0 +1,52 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::expr_or_init;
+use clippy_utils::source::snippet_opt;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+
+use super::JOIN_ABSOLUTE_PATHS;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, recv: &'tcx Expr<'tcx>, join_arg: &'tcx Expr<'tcx>, expr_span: Span) {
+    let ty = cx.typeck_results().expr_ty(recv).peel_refs();
+    if (is_type_diagnostic_item(cx, ty, sym::Path) || is_type_diagnostic_item(cx, ty, sym::PathBuf))
+        && let ExprKind::Lit(spanned) = expr_or_init(cx, join_arg).kind
+        && let LitKind::Str(symbol, _) = spanned.node
+        && let sym_str = symbol.as_str()
+        && (sym_str.starts_with('/') || sym_str.starts_with('\\'))
+    {
+        span_lint_and_then(
+            cx,
+            JOIN_ABSOLUTE_PATHS,
+            join_arg.span,
+            "argument to `Path::join` starts with a path separator",
+            |diag| {
+                let arg_str = snippet_opt(cx, spanned.span).unwrap_or_else(|| "..".to_string());
+
+                let no_separator = if sym_str.starts_with('/') {
+                    arg_str.replacen('/', "", 1)
+                } else {
+                    arg_str.replacen('\\', "", 1)
+                };
```

Comment:
Assume, you suggest this to safe performance? I feel like this refactoring would make the code less readable while having almost no impact. The closure is only called if the lint is actually emitted. I doubt that this would even be measurable.

I noticed another "optimization", of only calling `starts_with(&['/', '\\'])` instead of two separate times. But that rules out this suggestion, from what I can tell

Reply:
Not really performance, it's just that seeing a function two times with the same arguments hurts my feelings :cry:.  starts_with(&['/', '\\']) is fine though :) Please apply that change ₍˄·͈༝·͈˄₎ฅ˒˒

Reply:
Already done :D